### ISSUE TITLE
Elasticsearch histogram agg bug, query crashes with an AssertionError

### ISSUE THREAD

**Issue by Derya** [strapdata/elassandra#288]
Hi all. I'm getting this peculiar error when running an Elasticsearch query:
```
2019-05-07 22:59:52,538 ERROR [elasticsearch[172.17.0.2][search][T#17]] org.apache.cassandra.service.CassandraDaemon$2.uncaughtException(CassandraDaemon.java:231) Exception in thread Thread[elasticsearch[172.17.0.2][search][T#17],5,main]
java.lang.AssertionError: shards must return data sorted by key
	at org.elasticsearch.search.aggregations.bucket.histogram.InternalHistogram.reduceBuckets(InternalHistogram.java:342)
<plus a bunch more stuff>
```
The Elasticsearch query hangs actually, and doesn't return until it times out. But the above is what I see in the logs. 

Unfortunately I've been looking at this for a few hours now and I haven't been able to replicate it in any scenario much simpler than with my original fully-fledged ES mapping / cassandra table (more on that later). I can't quite share the whole thing but I can describe some aspects of it. The ES query in question looks something like this:

```
{
  "size": 0,
  "aggs": {
    "NEST": {
      "nested": {
        "path": "my_udt_col"
      },
      "aggs": {
        "X_COORD": {
          "histogram": {
            "interval": 0.1,
            "field": "my_udt_col.x_coord"
          },
          "aggs": {
            "Y_COORD": {
              "histogram": {
                "interval": 0.1,
                "field": "my_udt_col.y_coord"
              }
            }
          }
        }
      }
    }
  }
}
```

So, the cassandra table has a column that has the type `list<FROZEN <my_udt>>`. The ES index's mapping is 100% auto discovered from this table. Any other ES query I can think to run works, including per document stuff instead of aggregating stuff. The only way I can produce this error is by having that histogram aggregation nested within another histogram aggregation as I've pasted above.

However, doing this doesn't always cause this error. I eventually found that some rows (or ES docs, I guess) are problematic and some are not. If I constrain the ES query such that it's only operating on the nonproblematic rows, it works fine. I've spent a bit of time trying to find a pattern between the ES docs that cause this error and those that do not but to no avail.

Problematically, I can't get the query to crash AT ALL by inserting my own small documents/rows manually. All the documents/rows that cause a crash are generated by some other system we have and are really huge, which is part of what is making it so impossible for me to find a way to replicate this issue.

Software version: Elassandra docker image 6.2.3.11

**Comment by vroyer** [comment]
Elassandra does not change the elasticsearch code involved in aggregation, so the first thing to do is the reproduce this on Elassandra, and then test the same dataset on elasticsearch. If you cloud provide information to reproduce this issue, it would be very helpful.

**Comment by Derya** [comment]
Okay, interesting. So in theory I should have the same bug if I have the same document on a regular old elasticsearch cluster? I'll test that out, and I'll work on getting this to a reproducible state without sharing our entire schema. Thanks as always.

**Comment by Derya** [comment]
Update for this. I was never able to figure out the conditions necessary to cause this bug, but it was present in regular old elasticsearch too when I put in the same documents. It was at least a necessary condition to have two histogram aggregations (one nested within the other) with data with both 0 and -0 present in it, but other simpler docs I tried running such aggs on with both 0 and -0 data points didn't always crash it and I gave up eventually on figuring out what exactly was going on. It was easy to fix the problem for my own use case since I didn't care about distinguishing between 0 and -0 so I just mapped them all to 0. In any case, not really an elassandra issue.
